src/main/java/org/openjdk/jextract/json/JSONObjects.java

```diff
@@ -0,0 +1,826 @@
+                final int length = ja.size();
+                Object arr = Array.newInstance(elemCls, length);
+                for (int i = 0; i < length; i++) {
+                    Array.set(arr, i, JSONObjects.toObject(ja.get(i), elemCls));
+                }
+                return arr;
+            }
+        } else if (targetType instanceof ParameterizedType pt) {
+            if (pt.getRawType() instanceof Class cls) {
+                // List<T>, ArrayList<T> or Iterable<T>
+                if (isListOrIterable(cls)) {
+                    Type compType = getFirstTypeArgument(pt);
+                    return convertToList(ja, compType);
+                // Set<T>, HashMap<T> or LinkedHashMap<T>
+                } else if (isSet(cls)) {
+                    Type compType = getFirstTypeArgument(pt);
+                    return convertToSet(ja, compType);
+                // EnumSet<T>
+                } else if (cls == EnumSet.class) {
+                    return convertToEnumSet(ja, pt);
+                }
+            }
+        }
+
+        throw cannotConvert(ja, targetType);
+    }
+
+    // convert JSONNull value to a Java oject of the given Type
+    private static Object convertNull(JSONNull jn, Type targetType) {
+        if (targetType instanceof Class cls && cls.isPrimitive()) {
+            throw cannotConvert(jn, targetType);
+        } else if (targetType == OptionalInt.class) {
+            return OptionalInt.empty();
+        } else if (targetType == OptionalLong.class) {
+            return OptionalLong.empty();
+        } else if (targetType == OptionalDouble.class) {
+            return OptionalDouble.empty();
+        } else {
+            return null;
+        }
+    }
+
+    // convert the JSONBoolean value to a Java object of the given Type
+    private static Object convertBoolean(JSONBoolean jb, Type targetType) {
+        if (targetType == boolean.class ||
+            targetType == Boolean.class ||
+            targetType == Object.class) {
+            return jb.asBoolean();
+        } else {
+            throw cannotConvert(jb, targetType);
+        }
+    }
+
+    private static final Map<Class<?>, Function<JSONNumber, Object>> numberConverters =
+        new IdentityHashMap<>();
+
+    private static void addNumberConverter(Class<?> cls, Function<JSONNumber, Object> func) {
+        numberConverters.put(cls, func);
+    }
+
+    static {
+        addNumberConverter(byte.class, JSONNumber::asByte);
+        addNumberConverter(Byte.class, JSONNumber::asByte);
+        addNumberConverter(short.class, JSONNumber::asShort);
+        addNumberConverter(Short.class, JSONNumber::asShort);
+        addNumberConverter(int.class, JSONNumber::asInt);
+        addNumberConverter(Integer.class, JSONNumber::asInt);
+        addNumberConverter(long.class, JSONNumber::asLong);
+        addNumberConverter(Long.class, JSONNumber::asLong);
+        addNumberConverter(OptionalInt.class, jn -> OptionalInt.of(jn.asInt()));
```

Comment:
These last four seem odd

Reply:
You mean the OptionalInt part? Thats to have property that can be dropped. long vs int targets are supported (JSONNumber is always long - is converted to smaller ints by bounds check)